dali/operators/audio/preemphasis_filter_op.cc

```diff
@@ -0,0 +1,87 @@
+// Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "dali/operators/audio/preemphasis_filter_op.h"
+#include "dali/core/static_switch.h"
+
+namespace dali {
+
+
+DALI_SCHEMA(PreemphasisFilter)
+                .DocStr(R"code(This operator performs preemphasis filter on the input data.
+This filter in simple form can be expressed by the formula:
+`Y(t) = X(t) - X(t-1)*coeff`)code")
+                .NumInput(1)
+                .NumOutput(detail::kNumOutputs)
+                .AddOptionalArg(detail::kCoeff, R"code(Preemphasis coefficient `coeff`)code", 0.f)
+                .AddOptionalArg(detail::kDtype, R"code(Data type for the output)code", DALI_FLOAT);
+
+DALI_REGISTER_OPERATOR(PreemphasisFilter, PreemphasisFilterCpu, CPU);
+
+#define PREEMPH_TYPES (uint8_t, int8_t, uint16_t, int16_t, uint32_t, int32_t, uint64_t, int64_t, float, double)  // NOLINT
+
+bool PreemphasisFilterCpu::SetupImpl(std::vector<::dali::OutputDesc> &output_desc,
+                                     const workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  AcquireArguments(ws);
+  output_desc.resize(detail::kNumOutputs);
+  output_desc[0].shape = input.shape();
+  TYPE_SWITCH(output_type_, type2id, DType, PREEMPH_TYPES, (
+          {
+            TypeInfo type;
+            type.SetType<DType>(output_type_);
+            output_desc[0].type = type;
+          }
+  ), DALI_FAIL(make_string("Unsupported output type: ", output_type_)))  // NOLINT
+  return true;
+}
+
+
+void PreemphasisFilterCpu::RunImpl(workspace_t<CPUBackend> &ws) {
+  const auto &input = ws.template InputRef<CPUBackend>(0);
+  auto &output = ws.OutputRef<CPUBackend>(0);
+  auto &tp = ws.GetThreadPool();
+
+  for (int sample_id = 0; sample_id < batch_size_; ++sample_id) {
```

Comment:
I would move this loop to the most nested type_switch - for the readability.

Reply:
Done